### PR TITLE
chore: enable full test coverage for Windows

### PR DESCRIPTION
**Pull request by johnmaguire**
After an unsuccessful night debugging #92, I decided to create a benchmark to try to narrow down the cause of the slowness. Unfortunately, I ran into unexplained "Access denied" errors.

I ran the test suite on Windows and saw similar errors in many tests I read on Stack Overflow that AV can be an issue here, so I turned off Windows Defender RTP, to no avail. I think this may be related to the AppData issues mentioned here: https://github.com/garethgeorge/backrest/pull/200#issuecomment-2041206584

To see if this was an issue with my environment, I enabled tests on Github where I discovered the "Access denied" errors do not occur. I have resolved most of the remaining test failures in individual commits for your review.

I'm not sure if there was an actual bug with the `ls` command on Windows, but I received the following error during TestLs`:

```
restic_test.go:261: failed to list directory: command "C:\\Program Files (x86)/restic/restic-0.16.4.exe ls --json 0bca3b482c620f384eae2ca000f7d5ce212171049e5563dd16bc4591345407d7 C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestLs1936002979\\002 --no-cache" failed: exit status 1
    Process STDOUT:
    Fatal: All path filters must be absolute, starting with a forward slash '/'
```

I resolved it with a `toPathFilter` command for Windows.

---

However, there seems to be an issue with queue ordering. I am seeing occasional failures in the `TestTaskQueueOrdering` test, and persistent failures in `TestFuzzTaskQueue`. These functions looked a bit dense, so I figured I'd post the PR as-is and see if you had any thoughts on the queue.

---

This PR also enables running the full test suite in Github, to encourage keeping them green. Happy to revert that part if you simply want to improve the existing tests, and continue the current testing strategy.

### CODE REVIEW

File: pkg/restic/path_win.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"strings"
+)
+
+// toPathFilter converts a directory path to a path filter. Restic requires
+// these to be absolute, starting with a forward slash. On Windows, we convert
+// the drive letter to a single character and prepend a forward slash.
+func toPathFilter(path string) string {
+	cleanedPath := filepath.Clean(path)
+	// If clean results in an empty string, ignoring the volume, it returns "."
+	if cleanedPath != path+"." {
```
Comment (garethgeorge) [inline pkg/restic/path_win.go:17]: Hey, not sure I follow what this check is doing -- in general we'll never call toPathFilter with an input that that contains `/../` e.g. `/foo/../`. If we do run into that type of case, I think it'd be fine for that behavior to be unbefined.

File: pkg/restic/path_win.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
+	before, after, found := strings.Cut(path, string(filepath.Separator))
+	if !found {
+		// e.g. if path is "C:"
+		before = path
+		after = ""
+	}
+
+	if len(before) == 2 && before[1] == ':' && before[0] >= 'A' && before[0] <= 'Z' {
```
Comment (garethgeorge) [inline pkg/restic/path_win.go:28]: If I'm understanding correctly, the goal is to take a path that looks like `Z:/foo...` and convert it to something like `/Z/foo`?

I think we can simplify this a bit e.g.

```
func toRepoPath(path string) string {
   path := filepath.Clean(path) // not quite sure if we need this, we'll always call with absolute paths
   sepIdx := strings.Index(filepath.Separator)
   if sepIdx == -1 || sepIdx != 2 || path[1] != ':' {
      return path
   }
   return "/" + path[0] + "/" + path[3:]
}
```

also renamed to `toRepoPath` which maybe better represents what we're doing e.g. mapping a windows path format to a path as it's stored in the restic repo?

File: pkg/restic/restic.go [hunk 3]
```diff
@@ -306,6 +306,7 @@ func (r *Repo) ListDirectory(ctx context.Context, snapshot string, path string,
 		// an empty path can trigger very expensive operations (e.g. iterates all files in the snapshot)
 		return nil, nil, errors.New("path must not be empty")
 	}
+	path = toPathFilter(path)
```
Comment (garethgeorge) [inline pkg/restic/restic.go:309]: Hey, can you elaborate a bit on what motivated this change? Is this for test coverage or is there also an issue affecting browsing snapshots on windows?

**Comment by garethgeorge** [comment]
Hey, this is awesome! Thanks for tracking the windows breakages down and patching these, left a few questions about the changes in the restic pkg -- mostly wondering what you ran into / what motivated them (and saw some places where we can maybe simplify).

The task queue failures are interesting -- I remember seeing those and better coverage of (and an improved implementation of) the task queue is actually what's currently on my plate as I suspect there are some bugs in there related to https://github.com/garethgeorge/backrest/issues/180. It's possible a related issue is surfacing in the windows fuzztest. 

For now I'm happy to land this test enablement though, do you mind adding t.Skip() to those remaining failures for the time being? :)

Overall this is hugely appreciated -- thanks for contributing it!

File: pkg/restic/restic.go (hunk 3 above)
Comment (johnmaguire) [reply pkg/restic/restic.go:309]: Thanks for checking that on that! I was skeptical this was a real bug, since I hadn't seen errors browsing on Windows. Will update in a bit.

File: pkg/restic/path_win.go (hunk 1 above)
Comment (johnmaguire) [reply pkg/restic/path_win.go:17]: It's a weird one... this was for the `C:` test case. [`filepath.Clean`](https://pkg.go.dev/path/filepath#Clean) states that if the path is an empty string, it returns dot. Well... it doesn't include a drive volume in its test. So `C:` will return `C:.` It's odd, tbh: https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/path/filepath/path.go;l=101-103

We can probably ignore this edge case when moving this into the test.

**Comment by johnmaguire** [comment]
Should be good to go! I do wonder if the task queue tests might actually be a time resolution issue? I think in general, Linux has more precise time resolutions than on Windows. I have to run now, or I'd try changing the `Millisecond` reference to `Second` in them.

I skipped the two I've seen fail, but it's possible there are other flaky tests as well. Not sure if it'd help here, but in some of my projects, I enable `-race` detector in CI tests.

**Comment by garethgeorge** [comment]
Time resolution could definitely be at play I've gone ahead and reworked the task queues as a more modular implementation which is easier to cover with tests (and understand), the new coverage seems to pass on windows which is an improvement!

Enabling `-race` detector is a good idea, doing some local runs I'm seeing there are a couple read / write races in the IO handling when executing restic commands, I'll work through those and get some fixes out.

**Comment by garethgeorge** [comment]
Went ahead and submitted a patch for some minor races found by enabling `--race` , this was a great idea. Thanks.

**Comment by garethgeorge** [comment]
Hi John, I'm evaluating relicensing Backrest as AGPL-3.0 . Thanks again for all of the contributions you made early on in Backrest's launch -- I think your changes are some of the most lasting in the code base after a number of refactors :)

I'm wondering if you'd be willing to retroactively [sign a CLA](https://contributoragreements.org/query2form/?_replyto=garethgeorge97@gmail.com&_subject=Contributor%20License%20Agreement%20E-Signing&_body=Fill%20out%20the%20following%20form,%20then%20sign%20your%20initials%20to%20complete%20the%20Contributor%20License%20Agreement.&agreement-type[]=individual&agreement-type[]=entity&fullname=&title=&company=&email-address=&physical-address=&your-initials=&original-agreement=https://contributoragreements.org/u2s/2llymj8bjs&signed-agreement_s=https%3A%2F%2Fcontributoragreements.org%2Fu2s%2F2llymj8bjs%3Fyour-date%3D%40_time%26your-name%3D%40fullname%26your-title%3D%40title%26your-address%3D%40email-address%26your-patents%3D%40Patent-IDs-and-Country_t%26process-url%3D%40_processurl%26action%3Dsign-%40agreement-type%26%40u2s&_processurl=@processurl&_action[0]=https://contributoragreements.org/query2email/&_action[1]=https://contributoragreements.org/query2update/&_next=View%20More%20Contributor%20License%20Agreement%20Signers.&_success=Thank%20you%20for%20using%20contributoragreements.org.%20The%20agreement%20has%20been%20signed%20and%20sent%20via%20E-Mail%20and%20will%20not%20be%20stored.&_submit=Sign%20Your%20Contributor%20License%20Agreement.&_processurl=https%3A%2F%2Fcontributoragreements.org%2Fu2s%2F2lm52x4gui)?

I've become aware of a few SAAS looking at selling Backrest which goes philosophically against what I created it for (well they're welcome to -- but their platforms should be OSS and self-hostable re: the same freedoms Backrest provides!).
